app/assets/javascripts/discourse/app/controllers/topic.js

```diff
@@ -1292,6 +1294,35 @@ export default Controller.extend(bufferedProperty("model"), {
     this.model.destroy(this.currentUser);
   },
 
+  retryOnRateLimit(times, promise, topicId) {
+    const currentTopicId = this.get("model.id");
+    topicId = topicId || currentTopicId;
+    if (topicId !== currentTopicId || times <= 0) {
+      // we navigated to another topic, so skip
+      return;
+    }
+
+    promise().catch((e) => {
+      const xhr = e.jqXHR;
+      if (
+        xhr &&
+        xhr.status === 429 &&
+        xhr.responseJSON &&
+        xhr.responseJSON.extras &&
+        xhr.responseJSON.extras.wait_seconds
+      ) {
+        let waitSeconds = xhr.responseJSON.extras.wait_seconds;
+        if (waitSeconds < 5) {
+          waitSeconds = 5;
+        }
+
+        later(() => {
```

Comment:
Bear in mind that later could happen after topic controller has been destroyed.

A cleaner pattern would be to hold reference to later and cancel it in willDestroy I think: https://api.emberjs.com/ember/3.21/classes/Controller/methods/willDestroy?anchor=willDestroy

Reply:
Is the topic controller ever destroyed? I added some protection in case the topic id flips

Reply:
Yes AFAIK, it shouldn't be destroyed in normal workflow but there might be cases where it would, like tests maybe?  It's just like DOM listeners you don't absolutely always need to remove them, I just think it makes things clearer/cleaner to do it, and might avoid some complex to track bugs in the future.